src/sftp.rs

```diff
@@ -367,11 +370,32 @@ impl Sftp {
             Err(self.last_error())
         }
     }
+
+    /// Shuts the sftp connection down.
+    ///
+    /// This should not be called manually since the destructor takes care of it.
+    /// However, it is useful for async wrappers that take care of async destruction.
+    pub fn shutdown(&mut self) -> Result<(), Error> {
+        if !self.closed {
+            self.rc(unsafe { raw::libssh2_sftp_shutdown(self.raw) })?;
+            // set flag for destructor that we don't need to close the file anymore
+            self.closed = true;
+        }
+        Ok(())
+    }
 }
 
 impl Drop for Sftp {
     fn drop(&mut self) {
-        unsafe { assert_eq!(raw::libssh2_sftp_shutdown(self.raw), 0) }
+        if !self.closed {
+            let rc = loop {
```

Comment:
This loop makes me feel slightly anxious, as it feels like we might
introduce the potential to busy wait here.

What do you think about setting it to blocking mode while we shutdown in the context of `drop`?

Reply:
Just to make sure that I understand everything (together with what you suggested below):

In `drop`: 
```
if inner is some:
    set to blocking
    shutdown
    set back to previous blocking state
```

In `shutdown`:
```
if error:
    return error
else:
    take inner session
```

Is this what you meant?

Reply:
Yeah; I think you left out the actual shutdown call from the second case, so just for the sake of making sure we're on the same wavelength:

for shutdown:
* Attempt shutdown.
   * On error, propagate the error
   * Otherwise, take the inner option and discard it because it is no longer valid

Something like this:

```rust
fn shutdown(&mut self) -> Result<(), Error> {
   let raw = match self.inner.as_ref() {
         Some(inner) => inner.raw,
         None => return Err( <already shutdown>);
   };
   self.rc(unsafe { raw::libssh2_sftp_shutdown(raw) })?;
   self.inner.take();
   Ok(())
}
```